### ISSUE TITLE
[BUG] Why is `sample_log_prob` of shape `td.shape` unlike the other keys which are `[*td.shape, 1]`?

### ISSUE THREAD

**Issue by skandermoalla** [pytorch/rl#1243]
## Describe the bug

Building a `ProbabilisticActor` module with `return_log_prob=True` makes it write the log-probability of the distribution sample in a given tensordict.  However, unlike the other keys in the tensordict (e.g. `done`, `advantage`,  `value_target`, ...) which have shape `[*tensordict.shape, 1]`, it has shape `tensordict.shape`, i.e., without the additional last dimension.

This creates problems when keys are multiplied for example to compute a simple [REINFORCE loss](https://github.com/pytorch/rl/blob/e1d1874a7f3bd0143a1a07434929f4c58c010ead/torchrl/objectives/reinforce.py#L166).
Say you have a batch of dimension `torch.Size([n])`, then `tensordict["sample_log_prob"] * tensordict["advantage"]` will be a tensor of shape `torch.Size([n, n])` and taking the mean over it will be different from what one expects.

I also expect this inconsistency to cause other issues in the future.

## To Reproduce

Running
```python
import torch
from tensordict.nn import TensorDictModule, make_functional
from torchrl.envs.libs.gym import GymEnv
from torchrl.modules import (
    ProbabilisticActor,
    NormalParamWrapper,
    TanhNormal,
)

env = GymEnv("Pendulum-v1")
td = env.rollout(3)
module = NormalParamWrapper(torch.nn.Linear(3, env.action_spec.shape[-1] * 2))
tensordict_module = TensorDictModule(
    module, in_keys=["observation"], out_keys=["loc", "scale"]
)
td_module = ProbabilisticActor(
    module=tensordict_module,
    spec=env.action_spec,
    in_keys=["loc", "scale"],
    distribution_class=TanhNormal,
    return_log_prob=True,
)
params = make_functional(td_module)
td = td_module(td, params=params)
print(td)

reinforce_loss = td["sample_log_prob"] * td["next", "reward"]
print(f"REINFORCE loss has shape {reinforce_loss.shape}")


print(
    "Mean with wrong shapes",
    (torch.arange(3) * torch.arange(3).unsqueeze(1)).float().mean(),
)
print(
    "Mean with correct shapes",
    (torch.arange(3).unsqueeze(1) * torch.arange(3).unsqueeze(1)).float().mean(),
)

```
Gives
```bash
TensorDict(
    fields={
        action: Tensor(shape=torch.Size([3, 1]), device=cpu, dtype=torch.float32, is_shared=False),
        done: Tensor(shape=torch.Size([3, 1]), device=cpu, dtype=torch.bool, is_shared=False),
        loc: Tensor(shape=torch.Size([3, 1]), device=cpu, dtype=torch.float32, is_shared=False),
        next: TensorDict(
            fields={
                done: Tensor(shape=torch.Size([3, 1]), device=cpu, dtype=torch.bool, is_shared=False),
                observation: Tensor(shape=torch.Size([3, 3]), device=cpu, dtype=torch.float32, is_shared=False),
                reward: Tensor(shape=torch.Size([3, 1]), device=cpu, dtype=torch.float32, is_shared=False)},
            batch_size=torch.Size([3]),
            device=cpu,
            is_shared=False),
        observation: Tensor(shape=torch.Size([3, 3]), device=cpu, dtype=torch.float32, is_shared=False),
        sample_log_prob: Tensor(shape=torch.Size([3]), device=cpu, dtype=torch.float32, is_shared=False),
        scale: Tensor(shape=torch.Size([3, 1]), device=cpu, dtype=torch.float32, is_shared=False)},
    batch_size=torch.Size([3]),
    device=cpu,
    is_shared=False)
REINFORCE loss has shape torch.Size([3, 3])
Mean with wrong shapes tensor(1.)
Mean with correct shapes tensor(1.6667)
```

## Expected behavior

`sample_log_prob` should be written with dimension `[*tensordict.shape, 1]`

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/pytorch/rl/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by vmoens** [comment]
Thanks for raising this.
It's a good point. There is no reason other than "this is the right shape for the log prob", but I agree that it is inconsitent.
The reason why reward and done have a singleton in the end is because you will often have pieces of code like
```
value = network(obs)
some_other_val = reward + value
```
If reward does not have a singleton, this can easily lead to silent bugs because reward will be broadcast to the shape of value.
@matteobettini and others have raised this in the past and I can't really think of a proper way of addressing it.

In MARL settings your reward shape will follow the agent's configuration so the problem is less relevant.

Now the question is do we want log-prob to follow the same convention? After all in PPO we do sort of `log_prob * reward` (very wrongly put) so we'd like to avoid the same bug...

Happy to hear your thoughts on the shape that reward and done should have before discussing if log_prob should be reshaped too.

**Comment by skandermoalla** [comment]
> If reward does not have a singleton, this can easily lead to silent bugs because reward will be broadcast to the shape of value.

Indeed! And that's exactly what happened with `sample_log_prob` when multiplied by the `advantage`. It was a silent bug that I had quite some trouble finding. (With a batch of size 2048 the error between the correct and wrong mean is of order $10^5$).

I think the solution you came up with is good and it didn't seem unnatural at all when I started using TorchRL.
It's rather the opposite, it's `sample_log_prob` missing that dimension that feels unnatural.

I don't have an extensive knowledge of the design choices made in TorchRL, but I would push for `sample_log_prob` to follow the same convention yes!

**Comment by vmoens** [comment]
Ok good to know!
Thanks for the feedback, I'll see how I can fix this
